// cocobase/dist/cjs/core/core.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cocobase = void 0;
const utils_js_1 = require("../utils/utils.js");
const errors_js_1 = require("../types/errors.js");
const functions_js_1 = require("./functions.js");
const auth_js_1 = __importDefault(require("./auth.js"));
const websockets_js_1 = require("../realtime/websockets.js");
const multiplayer_js_1 = require("../realtime/multiplayer.js");
/**
 * Main Cocobase client for interacting with the Cocobase backend API.
 *
 * @example
 * ```typescript
 * const db = new Cocobase({
 *   apiKey: 'your-api-key',
 *   projectId: 'your-project-id',
 * });
 *
 * // Restore session on app load
 * const { isAuthenticated, user } = await db.auth.initAuth();
 *
 * // CRUD
 * await db.createDocument('posts', { title: 'Hello world' });
 * const posts = await db.listDocuments('posts', { filters: { published: true } });
 *
 * // Live queries (one-liner)
 * const unsub = db.onSnapshot('posts', (posts) => renderPosts(posts));
 *
 * // Pagination
 * const pager = db.paginate<Post>('posts', { sort: 'created_at' }, 20);
 * const { data, hasMore } = await pager.nextPage();
 * ```
 */
class Cocobase {
    constructor(config) {
        this.baseURL = config.baseURL ?? utils_js_1.BASEURL;
        this.apiKey = config.apiKey;
        this.projectId = config.projectId;
        this.timeout = config.timeout ?? 30000;
        this.auth = new auth_js_1.default(config);
        this.functions = new functions_js_1.CloudFunction(config.projectId || "project id required", () => this.auth.getToken());
        this.realtime = {
            collection: (collectionName, filters) => new websockets_js_1.CollectionWatcher(collectionName, this.apiKey || "", filters),
            broadcast: (userId, userName) => new websockets_js_1.ProjectBroadcast(this.apiKey || "", userId, userName),
            room: (roomId, userId, userName) => new websockets_js_1.RoomChat(roomId, this.apiKey || "", userId, userName),
            listRooms: async () => (0, websockets_js_1.listRooms)(this.apiKey || ""),
            game: (functionName) => new multiplayer_js_1.GameClient(this.projectId || "", functionName, this.auth.getToken()),
            listGameRooms: (publicOnly = true) => (0, multiplayer_js_1.listGameRooms)(this.projectId || "", this.apiKey, publicOnly),
        };
    }
    // ─── Internal request helper ──────────────────────────────────────────────────
    async request(method, path, body, useDataKey = true) {
        const url = `${this.baseURL}${path}`;
        const data = useDataKey ? { data: body } : body;
        const controller = new AbortController();
        let timer;
        if (this.timeout > 0) {
            timer = setTimeout(() => controller.abort(), this.timeout);
        }
        try {
            const res = await fetch(url, {
                method,
                headers: {
                    "Content-Type": "application/json",
                    ...(this.apiKey ? { "x-api-key": `${this.apiKey}` } : {}),
                    ...(this.auth.getToken()
                        ? { Authorization: `Bearer ${this.auth.getToken()}` }
                        : {}),
                },
                ...(body ? { body: JSON.stringify(data) } : {}),
                signal: controller.signal,
            });
            if (!res.ok) {
                const errorText = await res.text();
                let errorDetail;
                try {
                    errorDetail = JSON.parse(errorText);
                }
                catch {
                    errorDetail = errorText;
                }
                throw (0, errors_js_1.createHttpError)(res.status, url, method, errorDetail, this.getErrorSuggestion(res.status, method));
            }
            return res.json();
        }
        catch (error) {
            if (error instanceof Error && error.name === "AbortError") {
                throw new errors_js_1.TimeoutError(url, method, this.timeout);
            }
            throw error;
        }
        finally {
            if (timer !== undefined)
                clearTimeout(timer);
        }
    }
    getErrorSuggestion(status, method) {
        switch (status) {
            case 401:
                return "Check if your API key is valid and properly set";
            case 403:
                return "You don't have permission to perform this action. Verify your access rights";
            case 404:
                return "The requested resource was not found. Verify the path and ID are correct";
            case 405:
                return `The ${method} method is not allowed for this endpoint. Check the API documentation`;
            case 429:
                return "You've exceeded the rate limit. Please wait before making more requests";
            default:
                return "Check the API documentation and verify your request format";
        }
    }
    // ─── Document CRUD ────────────────────────────────────────────────────────────
    /**
     * Retrieves a single document by ID.
     *
     * @example
     * ```typescript
     * const post = await db.getDocument<Post>('posts', 'post-123');
     * console.log(post.data.title);
     * ```
     */
    async getDocument(collection, docId) {
        return this.request("GET", `/collections/${collection}/documents/${docId}`);
    }
    /**
     * Creates a new document in a collection.
     *
     * @example
     * ```typescript
     * const post = await db.createDocument('posts', { title: 'Hello', published: false });
     * ```
     */
    async createDocument(collection, data) {
        return this.request("POST", `/collections/documents?collection=${collection}`, data);
    }
    /**
     * Create a document with file uploads.
     *
     * @example
     * ```typescript
     * await db.createDocumentWithFiles('products',
     *   { name: 'Laptop', price: 1299 },
     *   { main_image: imageFile, gallery: [img1, img2] }
     * );
     * ```
     */
    async createDocumentWithFiles(collection, data, files) {
        const formData = new FormData();
        formData.append("data", JSON.stringify(data));
        for (const [fieldName, fileOrFiles] of Object.entries(files)) {
            if (Array.isArray(fileOrFiles)) {
                fileOrFiles.forEach((file) => formData.append(fieldName, file));
            }
            else {
                formData.append(fieldName, fileOrFiles);
            }
        }
        const url = `${this.baseURL}/collections/documents?collection=${collection}`;
        const res = await fetch(url, {
            method: "POST",
            headers: {
                ...(this.apiKey ? { "x-api-key": `${this.apiKey}` } : {}),
                ...(this.auth.getToken()
                    ? { Authorization: `Bearer ${this.auth.getToken()}` }
                    : {}),
            },
            body: formData,
        });
        if (!res.ok) {
            const errorText = await res.text();
            throw new Error(`File upload failed: ${errorText}`);
        }
        return res.json();
    }
    /**
     * Updates an existing document (partial patch — only provided fields are changed).
     *
     * @example
     * ```typescript
     * await db.updateDocument('posts', 'post-123', { published: true });
     * ```
     */
    async updateDocument(collection, docId, data) {
        return this.request("PATCH", `/collections/${collection}/documents/${docId}`, data);
    }
    /**
     * Replaces a document's data entirely (same as updateDocument but signals intent clearly).
     *
     * @example
     * ```typescript
     * await db.replaceDocument('settings', 'config-1', { theme: 'dark', lang: 'en' });
     * ```
     */
    async replaceDocument(collection, docId, data) {
        return this.request("PATCH", `/collections/${collection}/documents/${docId}`, data);
    }
    /**
     * Update a document with file uploads.
     *
     * @example
     * ```typescript
     * await db.updateDocumentWithFiles('users', 'user-123',
     *   { bio: 'Updated bio' },
     *   { avatar: newAvatarFile }
     * );
     * ```
     */
    async updateDocumentWithFiles(collection, docId, data, files) {
        const formData = new FormData();
        if (data) {
            formData.append("data", JSON.stringify(data));
        }
        if (files) {
            for (const [fieldName, fileOrFiles] of Object.entries(files)) {
                if (Array.isArray(fileOrFiles)) {
                    fileOrFiles.forEach((file) => formData.append(fieldName, file));
                }
                else {
                    formData.append(fieldName, fileOrFiles);
                }
            }
        }
        const url = `${this.baseURL}/collections/${collection}/documents/${docId}`;
        const res = await fetch(url, {
            method: "PATCH",
            headers: {
                ...(this.apiKey ? { "x-api-key": `${this.apiKey}` } : {}),
                ...(this.auth.getToken()
                    ? { Authorization: `Bearer ${this.auth.getToken()}` }
                    : {}),
            },
            body: formData,
        });
        if (!res.ok) {
            const errorText = await res.text();
            throw new Error(`File upload failed: ${errorText}`);
        }
        return res.json();
    }
    /**
     * Deletes a document from a collection.
     *
     * @example
     * ```typescript
     * await db.deleteDocument('posts', 'post-123');
     * ```
     */
    async deleteDocument(collection, docId) {
        return this.request("DELETE", `/collections/${collection}/documents/${docId}`);
    }
    /**
     * Lists documents with optional filtering, sorting, and pagination.
     *
     * @example
     * ```typescript
     * const posts = await db.listDocuments<Post>('posts', {
     *   filters: { published: true },
     *   sort: 'created_at',
     *   order: 'desc',
     *   limit: 20,
     * });
     * ```
     */
    async listDocuments(collection, query) {
        const query_str = (0, utils_js_1.buildFilterQuery)(query);
        return this.request("GET", `/collections/${collection}/documents${query_str ? `?${query_str}` : ""}`);
    }
    onSnapshot(collection, queryOrCallback, callback) {
        let query;
        let cb;
        if (typeof queryOrCallback === "function") {
            cb = queryOrCallback;
        }
        else {
            query = queryOrCallback;
            cb = callback;
        }
        const realtimeFilters = query?.filters;
        const watcher = new websockets_js_1.CollectionWatcher(collection, this.apiKey || "", realtimeFilters);
        const refresh = () => {
            this.listDocuments(collection, query)
                .then(cb)
                .catch(() => { });
        };
        // Initial fetch
        refresh();
        watcher.connect();
        watcher.onCreate(() => refresh());
        watcher.onUpdate(() => refresh());
        watcher.onDelete(() => refresh());
        return () => watcher.disconnect();
    }
    // ─── Pagination ───────────────────────────────────────────────────────────────
    /**
     * Creates a paginator for a collection that tracks page state client-side.
     *
     * @param collection - Collection name
     * @param query - Optional base query (filters, sort, etc.) — do not set `limit`/`offset` here
     * @param pageSize - Number of documents per page (default: 20)
     *
     * @example
     * ```typescript
     * const pager = db.paginate<Post>('posts', { sort: 'created_at', order: 'desc' }, 20);
     *
     * const page1 = await pager.nextPage();  // { data, hasMore, page: 0, pageSize: 20, offset: 0 }
     * const page2 = await pager.nextPage();  // { data, hasMore, page: 1, pageSize: 20, offset: 20 }
     *
     * if (page2.hasMore) {
     *   const page3 = await pager.nextPage();
     * }
     *
     * await pager.prevPage();       // go back one page
     * await pager.goToPage(0);      // jump to first page
     * pager.reset();                // reset to beginning without fetching
     * ```
     */
    paginate(collection, query, pageSize = 20) {
        let currentPage = -1;
        let done = false;
        const fetchPage = async (page) => {
            const offset = page * pageSize;
            const q = { ...query, limit: pageSize + 1, offset };
            const results = await this.listDocuments(collection, q);
            const hasMore = results.length > pageSize;
            const data = hasMore ? results.slice(0, pageSize) : results;
            currentPage = page;
            done = !hasMore;
            return { data, hasMore, page, pageSize, offset };
        };
        return {
            /** Fetch the next page. */
            nextPage: () => fetchPage(currentPage + 1),
            /** Fetch the previous page. Returns empty if already on page 0. */
            prevPage: () => {
                if (currentPage <= 0)
                    return fetchPage(0);
                return fetchPage(currentPage - 1);
            },
            /** Jump to a specific page index (0-based). */
            goToPage: (page) => fetchPage(Math.max(0, page)),
            /** Whether the last fetched page had a next page. */
            get hasMore() {
                return !done;
            },
            /** Current 0-based page index (-1 if no page fetched yet). */
            get currentPage() {
                return currentPage;
            },
            /** Reset pagination to start (does not fetch). */
            reset() {
                currentPage = -1;
                done = false;
            },
        };
    }
    // ─── Batch operations ─────────────────────────────────────────────────────────
    /**
     * Delete multiple documents by ID.
     *
     * @example
     * ```typescript
     * await db.deleteDocuments('posts', ['id1', 'id2', 'id3']);
     * ```
     */
    async deleteDocuments(collection, docIds) {
        return this.request("POST", `/collections/${collection}/batch/documents/delete`, { document_ids: docIds }, false);
    }
    /**
     * Create multiple documents in a single request.
     *
     * @example
     * ```typescript
     * const created = await db.createDocuments('posts', [
     *   { title: 'Post 1' },
     *   { title: 'Post 2' },
     * ]);
     * ```
     */
    async createDocuments(collection, documents) {
        return this.request("POST", `/collections/${collection}/batch/documents/create`, { documents }, false);
    }
    /**
     * Update multiple documents in a single request.
     *
     * @param updates - Map of document ID → partial update data
     *
     * @example
     * ```typescript
     * await db.updateDocuments('posts', {
     *   'post-1': { published: true },
     *   'post-2': { published: true },
     * });
     * ```
     */
    async updateDocuments(collection, updates) {
        return this.request("POST", `/collections/${collection}/batch/documents/update`, { updates }, false);
    }
    // ─── Query helpers ────────────────────────────────────────────────────────────
    /**
     * Count documents matching a query without fetching them.
     *
     * @example
     * ```typescript
     * const { count } = await db.countDocuments('users', { filters: { active: true } });
     * ```
     */
    async countDocuments(collection, query) {
        const query_str = (0, utils_js_1.buildFilterQuery)(query);
        return this.request("GET", `/collections/${collection}/query/documents/count${query_str ? `?${query_str}` : ""}`);
    }
    /**
     * Aggregate documents (count, sum, avg, min, max).
     *
     * @example
     * ```typescript
     * const { result } = await db.aggregateDocuments('orders', {
     *   field: 'amount',
     *   operation: 'sum',
     *   query: { filters: { status: 'completed' } },
     * });
     * console.log('Total revenue:', result);
     * ```
     */
    async aggregateDocuments(collection, params) {
        const query_str = (0, utils_js_1.buildFilterQuery)(params.query);
        return (await this.request("GET", `/collections/${collection}/query/documents/aggregate?field=${params.field}&operation=${params.operation}&${query_str ? `${query_str}` : ""}`));
    }
    // ─── Deprecated shims ─────────────────────────────────────────────────────────
    /** @deprecated Use `db.auth.initAuth()` instead. */
    async initAuth(token) {
        return this.auth.initAuth(token);
    }
    /** @deprecated Use `db.auth.login()` instead. */
    async login(params) {
        return this.auth.login(params);
    }
    /** @deprecated Use `db.auth.register()` instead. */
    async register(params) {
        return this.auth.register(params);
    }
    /** @deprecated Use `db.auth.loginWithGoogle()` instead. */
    async loginWithGoogle(params) {
        return this.auth.loginWithGoogle(params);
    }
    /** @deprecated Use `db.auth.registerWithFiles()` instead. */
    async registerWithFiles(params) {
        return this.auth.registerWithFiles(params);
    }
    /** @deprecated Use `db.auth.logout()` instead. */
    logout() {
        this.auth.logout();
    }
    /** @deprecated Use `db.auth.isAuthenticated()` instead. */
    isAuthenticated() {
        return this.auth.isAuthenticated();
    }
    /** @deprecated Use `db.auth.getCurrentUser()` instead. */
    async getCurrentUser() {
        return this.auth.getCurrentUser();
    }
}
exports.Cocobase = Cocobase;
