// cocobase/dist/cjs/core/core.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cocobase = void 0;
const utils_js_1 = require("../utils/utils.js");
const functions_js_1 = require("./functions.js");
const auth_js_1 = __importDefault(require("./auth.js"));
const websockets_js_1 = require("../realtime/websockets.js");
const multiplayer_js_1 = require("../realtime/multiplayer.js");
/**
 * Main Cocobase client for interacting with the Cocobase backend API.
 *
 * Provides methods for:
 * - Document CRUD operations (Create, Read, Update, Delete)
 * - User authentication and management
 * - File uploads
 * - Real-time data synchronization
 * - Cloud functions execution
 * - Batch operations
 * - Advanced querying and aggregations
 *
 * @example
 * ```typescript
 * // Initialize the client
 * const db = new Cocobase({
 *   apiKey: 'your-api-key',
 *   projectId: 'your-project-id'
 * });
 *
 * // Create a document
 * await db.createDocument('users', { name: 'John Doe' });
 *
 * // Query documents
 * const users = await db.listDocuments('users', {
 *   filters: { status: 'active' },
 *   limit: 10
 * });
 * ```
 */
class Cocobase {
    /**
     * Creates a new Cocobase client instance.
     *
     * @param config - Configuration object for the client
     * @param config.apiKey - Your Cocobase API key (required for most operations)
     * @param config.projectId - Your Cocobase project ID (required for cloud functions)
     * @param config.baseURL - Optional custom base URL (defaults to https://api.cocobase.buzz)
     *
     * @example
     * ```typescript
     * const db = new Cocobase({
     *   apiKey: 'your-api-key',
     *   projectId: 'your-project-id'
     * });
     * ```
     */
    constructor(config) {
        this.baseURL = config.baseURL ?? utils_js_1.BASEURL;
        this.apiKey = config.apiKey;
        this.projectId = config.projectId;
        this.auth = new auth_js_1.default(config);
        this.functions = new functions_js_1.CloudFunction(config.projectId || "project id required", () => this.auth.getToken());
        // realtime factories bound to this client configuration
        this.realtime = {
            collection: (collectionName, filters) => new websockets_js_1.CollectionWatcher(collectionName, this.apiKey || "", filters),
            broadcast: (userId, userName) => new websockets_js_1.ProjectBroadcast(this.apiKey || "", userId, userName),
            room: (roomId, userId, userName) => new websockets_js_1.RoomChat(roomId, this.apiKey || "", userId, userName),
            listRooms: async () => (0, websockets_js_1.listRooms)(this.apiKey || ""),
            game: (functionName) => new multiplayer_js_1.GameClient(this.projectId || "", functionName, this.auth.getToken()),
            listGameRooms: (publicOnly = true) => (0, multiplayer_js_1.listGameRooms)(this.projectId || "", this.apiKey, publicOnly),
        };
    }
    async request(method, path, body, useDataKey = true) {
        const url = `${this.baseURL}${path}`;
        const data = useDataKey ? { data: body } : body;
        try {
            const res = await fetch(url, {
                method,
                headers: {
                    "Content-Type": "application/json",
                    ...(this.apiKey ? { "x-api-key": `${this.apiKey}` } : {}),
                    ...(this.auth.getToken()
                        ? { Authorization: `Bearer ${this.auth.getToken()}` }
                        : {}),
                },
                ...(body ? { body: JSON.stringify(data) } : {}),
            });
            if (!res.ok) {
                const errorText = await res.text();
                let errorDetail;
                try {
                    errorDetail = JSON.parse(errorText);
                }
                catch {
                    errorDetail = errorText;
                }
                const errorMessage = {
                    statusCode: res.status,
                    url,
                    method,
                    error: errorDetail,
                    suggestions: this.getErrorSuggestion(res.status, method),
                };
                throw new Error(`Request failed:\n${JSON.stringify(errorMessage, null, 2)}`);
            }
            return res.json();
        }
        catch (error) {
            if (error instanceof Error) {
                throw error;
            }
            throw new Error(`Unexpected error during ${method} request to ${url}: ${error}`);
        }
    }
    getErrorSuggestion(status, method) {
        switch (status) {
            case 401:
                return "Check if your API key is valid and properly set";
            case 403:
                return "You don't have permission to perform this action. Verify your access rights";
            case 404:
                return "The requested resource was not found. Verify the path and ID are correct";
            case 405:
                return `The ${method} method is not allowed for this endpoint. Check the API documentation for supported methods`;
            case 429:
                return "You've exceeded the rate limit. Please wait before making more requests";
            default:
                return "Check the API documentation and verify your request format";
        }
    }
    /**
     * Retrieves a single document by ID from a collection.
     *
     * @template T - The type of the document data
     * @param collection - Name of the collection
     * @param docId - Unique ID of the document
     * @returns Promise resolving to the document with metadata
     *
     * @example
     * ```typescript
     * const user = await db.getDocument('users', 'user-123');
     * console.log(user.data.name);
     * ```
     */
    async getDocument(collection, docId) {
        return this.request("GET", `/collections/${collection}/documents/${docId}`);
    }
    /**
     * Creates a new document in a collection.
     *
     * @template T - The type of the document data
     * @param collection - Name of the collection
     * @param data - Document data to store
     * @returns Promise resolving to the created document with metadata
     *
     * @example
     * ```typescript
     * const newUser = await db.createDocument('users', {
     *   name: 'John Doe',
     *   email: 'john@example.com',
     *   age: 30
     * });
     * ```
     */
    async createDocument(collection, data) {
        return this.request("POST", `/collections/documents?collection=${collection}`, data);
    }
    /**
     * Create a document with file uploads
     *
     * @param collection - Collection name
     * @param data - Document data (JSON object)
     * @param files - Object mapping field names to File objects
     *
     * @example
     * ```typescript
     * // Upload avatar and cover photo
     * await db.createDocumentWithFiles('users',
     *   { name: 'John Doe', email: 'john@example.com' },
     *   {
     *     avatar: avatarFile,
     *     cover_photo: coverFile
     *   }
     * );
     *
     * // Upload product with gallery
     * await db.createDocumentWithFiles('products',
     *   { name: 'Laptop', price: 1299 },
     *   {
     *     main_image: mainImageFile,
     *     gallery: [img1, img2, img3] // Array for multiple files
     *   }
     * );
     * ```
     */
    async createDocumentWithFiles(collection, data, files) {
        const formData = new FormData();
        // Add JSON data
        formData.append("data", JSON.stringify(data));
        // Add files with their field names
        for (const [fieldName, fileOrFiles] of Object.entries(files)) {
            if (Array.isArray(fileOrFiles)) {
                // Multiple files with same field name creates an array
                fileOrFiles.forEach((file) => {
                    formData.append(fieldName, file);
                });
            }
            else {
                // Single file
                formData.append(fieldName, fileOrFiles);
            }
        }
        const url = `${this.baseURL}/collections/documents?collection=${collection}`;
        const res = await fetch(url, {
            method: "POST",
            headers: {
                ...(this.apiKey ? { "x-api-key": `${this.apiKey}` } : {}),
                ...(this.auth.getToken()
                    ? { Authorization: `Bearer ${this.auth.getToken()}` }
                    : {}),
            },
            body: formData,
        });
        if (!res.ok) {
            const errorText = await res.text();
            throw new Error(`File upload failed: ${errorText}`);
        }
        return res.json();
    }
    /**
     * Updates an existing document in a collection.
     *
     * @template T - The type of the document data
     * @param collection - Name of the collection
     * @param docId - Unique ID of the document to update
     * @param data - Partial document data to update (only specified fields are updated)
     * @returns Promise resolving to the updated document with metadata
     *
     * @example
     * ```typescript
     * await db.updateDocument('users', 'user-123', {
     *   age: 31,
     *   status: 'active'
     * });
     * ```
     */
    async updateDocument(collection, docId, data) {
        return this.request("PATCH", `/collections/${collection}/documents/${docId}`, data);
    }
    /**
     * Update a document with file uploads
     *
     * @param collection - Collection name
     * @param docId - Document ID
     * @param data - Partial document data to update (optional)
     * @param files - Object mapping field names to File objects (optional)
     *
     * @example
     * ```typescript
     * // Update only avatar
     * await db.updateDocumentWithFiles('users', 'user-123',
     *   undefined,
     *   { avatar: newAvatarFile }
     * );
     *
     * // Update data and files
     * await db.updateDocumentWithFiles('users', 'user-123',
     *   { bio: 'Updated bio' },
     *   { avatar: newAvatarFile, cover_photo: newCoverFile }
     * );
     * ```
     */
    async updateDocumentWithFiles(collection, docId, data, files) {
        const formData = new FormData();
        // Add JSON data if provided
        if (data) {
            formData.append("data", JSON.stringify(data));
        }
        // Add files with their field names if provided
        if (files) {
            for (const [fieldName, fileOrFiles] of Object.entries(files)) {
                if (Array.isArray(fileOrFiles)) {
                    fileOrFiles.forEach((file) => {
                        formData.append(fieldName, file);
                    });
                }
                else {
                    formData.append(fieldName, fileOrFiles);
                }
            }
        }
        const url = `${this.baseURL}/collections/${collection}/documents/${docId}`;
        const res = await fetch(url, {
            method: "PATCH",
            headers: {
                ...(this.apiKey ? { "x-api-key": `${this.apiKey}` } : {}),
                ...(this.auth.getToken()
                    ? { Authorization: `Bearer ${this.auth.getToken()}` }
                    : {}),
            },
            body: formData,
        });
        if (!res.ok) {
            const errorText = await res.text();
            throw new Error(`File upload failed: ${errorText}`);
        }
        return res.json();
    }
    /**
     * Deletes a document from a collection.
     *
     * @param collection - Name of the collection
     * @param docId - Unique ID of the document to delete
     * @returns Promise resolving to a success status object
     *
     * @example
     * ```typescript
     * await db.deleteDocument('users', 'user-123');
     * ```
     */
    async deleteDocument(collection, docId) {
        return this.request("DELETE", `/collections/${collection}/documents/${docId}`);
    }
    /**
     * Lists documents from a collection with optional filtering and pagination.
     *
     * @template T - The type of the document data
     * @param collection - Name of the collection
     * @param query - Optional query parameters for filtering, sorting, and pagination
     * @returns Promise resolving to an array of documents
     *
     * @example
     * ```typescript
     * // Simple query
     * const users = await db.listDocuments('users', {
     *   filters: { status: 'active' },
     *   limit: 10
     * });
     *
     * // Advanced query with sorting
     * const posts = await db.listDocuments('posts', {
     *   filters: { published: true },
     *   sort: 'createdAt',
     *   order: 'desc',
     *   limit: 20
     * });
     * ```
     */
    async listDocuments(collection, query) {
        const query_str = (0, utils_js_1.buildFilterQuery)(query);
        return this.request("GET", `/collections/${collection}/documents${query_str ? `?${query_str}` : ""}`);
    }
    /**
     * Initializes authentication by restoring the session from local storage.
     * Call this method when your application loads to restore user sessions.
     *
     * @deprecated Use `db.auth.initAuth()` instead. This method will be removed in a future version.
     * @example
     * ```typescript
     * await db.initAuth();
     * if (db.isAuthenticated()) {
     *   console.log('User is logged in:', db.user);
     * }
     * ```
     */
    async initAuth() {
        return this.auth.initAuth();
    }
    /**
     * Authenticates a user with email and password.
     *
     * @deprecated Use `db.auth.login()` instead. This method will be removed in a future version.
     * @param params - Login parameters
     * @returns Promise that resolves when login is complete
     *
     * @example
     * ```typescript
     * await db.login({ email: 'user@example.com', password: 'password123' });
     * console.log('Logged in as:', db.user.email);
     * ```
     */
    async login(params) {
        return this.auth.login(params);
    }
    /**
     * Registers a new user with email, password, and optional additional data.
     *
     * @deprecated Use `db.auth.register()` instead. This method will be removed in a future version.
     * @param params - Registration parameters
     * @returns Promise that resolves when registration is complete
     *
     * @example
     * ```typescript
     * await db.register({
     *   email: 'user@example.com',
     *   password: 'password123',
     *   data: { username: 'johndoe', fullName: 'John Doe' }
     * });
     * ```
     */
    async register(params) {
        return this.auth.register(params);
    }
    /**
     * Authenticates a user using Google Sign-In with ID token.
     *
     * @deprecated Use `db.auth.loginWithGoogle()` instead. This method will be removed in a future version.
     * @param params - Google login parameters
     * @returns Promise resolving to the authenticated user object
     *
     * @example
     * ```typescript
     * // New recommended way
     * const user = await db.auth.loginWithGoogle({ idToken, platform: 'web' });
     *
     * // Old way (deprecated)
     * const user = await db.loginWithGoogle({ idToken, platform: 'web' });
     * ```
     */
    async loginWithGoogle(params) {
        return this.auth.loginWithGoogle(params);
    }
    /**
     * Register a new user with file uploads (avatar, cover photo, etc.)
     *
     * @deprecated Use `db.auth.registerWithFiles()` instead. This method will be removed in a future version.
     * @param params - Registration parameters with files
     *
     * @example
     * ```typescript
     * // Register with avatar
     * await db.registerWithFiles({
     *   email: 'john@example.com',
     *   password: 'password123',
     *   data: { username: 'johndoe', full_name: 'John Doe' },
     *   files: { avatar: avatarFile }
     * });
     *
     * // Register with avatar and cover photo
     * await db.registerWithFiles({
     *   email: 'john@example.com',
     *   password: 'password123',
     *   data: { username: 'johndoe' },
     *   files: { avatar: avatarFile, cover_photo: coverFile }
     * });
     * ```
     */
    async registerWithFiles(params) {
        return this.auth.registerWithFiles(params);
    }
    /**
     * Logs out the current user by clearing the token and user data.
     *
     * @deprecated Use `db.auth.logout()` instead. This method will be removed in a future version.
     *
     * @example
     * ```typescript
     * db.logout();
     * console.log('User logged out');
     * ```
     */
    logout() {
        this.auth.logout();
    }
    /**
     * Checks if a user is currently authenticated.
     *
     * @deprecated Use `db.auth.isAuthenticated()` instead. This method will be removed in a future version.
     * @returns True if user is authenticated, false otherwise
     *
     * @example
     * ```typescript
     * if (db.isAuthenticated()) {
     *   console.log('User is logged in');
     * }
     * ```
     */
    isAuthenticated() {
        return this.auth.isAuthenticated();
    }
    /**
     * @deprecated Use `db.auth.getCurrentUser()` instead. This method will be removed in a future version.
     */
    async getCurrentUser() {
        return this.auth.getCurrentUser();
    }
    // BATCH OPERATIONS
    async deleteDocuments(collection, docIds) {
        return this.request("POST", `/collections/${collection}/batch/documents/delete`, { document_ids: docIds }, false);
    }
    async createDocuments(collection, documents) {
        return this.request("POST", `/collections/${collection}/batch/documents/create`, { documents }, false);
    }
    /**
     * Batch update documents
     *
     * @param collection - Collection name
     * @param updates - Object mapping document IDs to partial update objects.
     *   Example: { "docId1": { fieldA: "value" }, "docId2": { fieldB: 2 } }
     */
    async updateDocuments(collection, updates) {
        return this.request("POST", `/collections/${collection}/batch/documents/update`, { updates }, false);
    }
    /**
     * Count documents matching filters without returning the documents.
     *
     * Example:
     *  await db.countDocuments('users', { status: 'active', age_gte: 18 })
     *  // returns { count: 42 }
     */
    async countDocuments(collection, query) {
        const query_str = (0, utils_js_1.buildFilterQuery)(query);
        return this.request("GET", `/collections/${collection}/query/documents/count${query_str ? `?${query_str}` : ""}`);
    }
    async aggregateDocuments(collection, params) {
        const query_str = (0, utils_js_1.buildFilterQuery)(params.query);
        return (await this.request("GET", `/collections/${collection}/query/documents/aggregate?field=${params.field}&operation=${params.operation}&${query_str ? `${query_str}` : ""}`));
    }
}
exports.Cocobase = Cocobase;
