// cocobase/dist/cjs/core/functions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudFunction = void 0;
const BASEURL = "https://cloud.cocobase.buzz";
/**
 * CloudFunction client for executing server-side functions.
 *
 * Cloud functions allow you to run custom server-side logic without managing infrastructure.
 * Functions are written in Python and deployed to your Cocobase project.
 *
 * @example
 * ```typescript
 * // Access via the Cocobase instance
 * const db = new Cocobase({
 *   apiKey: 'your-api-key',
 *   projectId: 'your-project-id'
 * });
 *
 * // Execute a function
 * const result = await db.functions.execute('sendEmail', {
 *   payload: { to: 'user@example.com', subject: 'Hello' },
 *   method: 'POST'
 * });
 * ```
 */
class CloudFunction {
    /**
     * Creates a new CloudFunction client.
     *
     * @param projectId - Your Cocobase project ID
     * @param getToken - Function that returns the current authentication token
     * @throws Error if projectId is empty or invalid
     */
    constructor(projectId, getToken) {
        this.projectId = projectId;
        this.getToken = getToken;
        // Validate projectId
        if (!projectId || projectId.trim() === "") {
            throw new Error("CloudFunction requires a valid projectId. Please provide projectId in CocobaseConfig.");
        }
    }
    /**
     * Executes a cloud function by name.
     *
     * @template T - The expected return type of the function
     * @param functionName - Name of the cloud function to execute
     * @param params - Optional parameters including payload and HTTP method
     * @returns Promise resolving to the function response with result and metadata
     * @throws Error if projectId is invalid
     *
     * @example
     * ```typescript
     * // Simple GET request
     * const result = await db.functions.execute('getStats');
     *
     * // POST request with payload
     * const result = await db.functions.execute('processOrder', {
     *   payload: {
     *     orderId: '12345',
     *     items: [{ id: 1, quantity: 2 }]
     *   },
     *   method: 'POST'
     * });
     *
     * console.log(result.result); // Function output
     * console.log(result.execution_time); // Execution time in ms
     * ```
     */
    async execute(functionName, params) {
        // Validate projectId again in case it was modified
        if (!this.projectId || this.projectId.trim() === "") {
            throw new Error("Invalid projectId. Please ensure projectId is set in CocobaseConfig.");
        }
        const url = `${BASEURL}/functions/${this.projectId}/func/${functionName}`;
        // Default to GET if no params provided, otherwise use specified method or POST
        const method = params?.method ?? (params?.payload ? "POST" : "GET");
        const fetchOptions = {
            method,
            headers: {
                "Content-Type": "application/json",
            },
        };
        // Get the latest token dynamically
        const token = this.getToken();
        if (token) {
            fetchOptions.headers = {
                ...fetchOptions.headers,
                Authorization: `Bearer ${token}`,
            };
        }
        // Only add body for POST/PUT requests with payload
        if (method === "POST" && params?.payload) {
            fetchOptions.body = JSON.stringify({
                payload: params.payload,
            });
        }
        const res = await fetch(url, fetchOptions);
        const data = await res.json();
        return data;
    }
}
exports.CloudFunction = CloudFunction;
