// cocobase/dist/core/auth.js
import { buildFilterQuery, mergeUserData } from "../utils/utils.js";
import { autoStorageAdapter } from "../utils/storage.js";
import { createHttpError, TimeoutError, UnauthorizedError, ForbiddenError, } from "../types/errors.js";
const TOKEN_KEY = "cocobase-token";
const USER_KEY = "cocobase-user";
/**
 * Authentication handler for Cocobase client.
 *
 * @example
 * ```typescript
 * const db = new Cocobase({ apiKey: 'your-key' });
 *
 * // Subscribe to auth state (returns unsubscribe)
 * const unsub = db.auth.onAuthStateChange((user, token) => {
 *   setUser(user ?? null);
 * });
 *
 * // Restore session on app load
 * const { isAuthenticated } = await db.auth.initAuth();
 * ```
 */
export class AuthHandler {
    constructor(config) {
        this.callbacks = {};
        this.authStateListeners = new Set();
        this.initAuthDone = false;
        this.waitForAuthResolvers = [];
        this.baseURL = config.baseURL ?? "https://api.cocobase.cc";
        this.apiKey = config.apiKey;
        this.storage = config.storage ?? autoStorageAdapter();
        this.timeout = config.timeout ?? 30000;
    }
    // ─── Internal helpers ────────────────────────────────────────────────────────
    async request(method, path, body, useDataKey = true) {
        const url = `${this.baseURL}${path}`;
        const data = useDataKey ? { data: body } : body;
        const controller = new AbortController();
        let timer;
        if (this.timeout > 0) {
            timer = setTimeout(() => controller.abort(), this.timeout);
        }
        try {
            const res = await fetch(url, {
                method,
                headers: {
                    "Content-Type": "application/json",
                    ...(this.apiKey ? { "x-api-key": `${this.apiKey}` } : {}),
                    ...(this.token ? { Authorization: `Bearer ${this.token}` } : {}),
                },
                ...(body ? { body: JSON.stringify(data) } : {}),
                signal: controller.signal,
            });
            if (!res.ok) {
                const errorText = await res.text();
                let errorDetail;
                try {
                    errorDetail = JSON.parse(errorText);
                }
                catch {
                    errorDetail = errorText;
                }
                throw createHttpError(res.status, url, method, errorDetail, this.getErrorSuggestion(res.status, method));
            }
            return res.json();
        }
        catch (error) {
            if (error instanceof Error && error.name === "AbortError") {
                throw new TimeoutError(url, method, this.timeout);
            }
            throw error;
        }
        finally {
            if (timer !== undefined)
                clearTimeout(timer);
        }
    }
    getErrorSuggestion(status, method) {
        switch (status) {
            case 401:
                return "Check if your API key is valid and properly set";
            case 403:
                return "You don't have permission to perform this action. Verify your access rights";
            case 404:
                return "The requested resource was not found. Verify the path and ID are correct";
            case 405:
                return `The ${method} method is not allowed for this endpoint. Check the API documentation`;
            case 429:
                return "You've exceeded the rate limit. Please wait before making more requests";
            default:
                return "Check the API documentation and verify your request format";
        }
    }
    decodeJwtPayload(token) {
        try {
            const base64 = token.split(".")[1];
            if (!base64)
                return null;
            const json = atob(base64.replace(/-/g, "+").replace(/_/g, "/"));
            return JSON.parse(json);
        }
        catch {
            return null;
        }
    }
    fireAuthStateChange() {
        this.callbacks.onAuthStateChange?.(this.user, this.token);
        this.authStateListeners.forEach((cb) => cb(this.user, this.token));
    }
    // ─── Token / user persistence ────────────────────────────────────────────────
    getToken() {
        return this.token;
    }
    setToken(token) {
        this.token = token;
        this.storage.setItem(TOKEN_KEY, token);
        this.callbacks.onTokenChange?.(token);
    }
    setUser(user) {
        this.user = user;
        this.storage.setItem(USER_KEY, JSON.stringify(user));
    }
    getUser() {
        return this.user;
    }
    isAuthenticated() {
        return !!this.token;
    }
    // ─── Token expiry helpers ─────────────────────────────────────────────────────
    /**
     * Returns the expiry date of the current JWT token, or null if unavailable.
     *
     * @example
     * ```typescript
     * const exp = db.auth.getTokenExpiry();
     * if (exp) console.log('Token expires:', exp.toLocaleString());
     * ```
     */
    getTokenExpiry() {
        if (!this.token)
            return null;
        const payload = this.decodeJwtPayload(this.token);
        if (!payload?.exp)
            return null;
        return new Date(payload.exp * 1000);
    }
    /**
     * Returns true if the current token has expired (based on JWT `exp` claim).
     * Returns false if no token, or if the token has no `exp` claim.
     */
    isTokenExpired() {
        if (!this.token)
            return false;
        const payload = this.decodeJwtPayload(this.token);
        if (!payload?.exp)
            return false;
        return Date.now() / 1000 > payload.exp;
    }
    // ─── Auth state subscription ──────────────────────────────────────────────────
    /**
     * Subscribe to auth state changes (login, logout, token change, initAuth).
     * Returns an unsubscribe function — call it to remove the listener.
     *
     * Mirrors Firebase's `onAuthStateChanged` API.
     *
     * @example
     * ```typescript
     * const unsub = db.auth.onAuthStateChange((user, token) => {
     *   setUser(user ?? null);
     *   setToken(token ?? null);
     * });
     *
     * // In React: call unsub in useEffect cleanup
     * useEffect(() => {
     *   return db.auth.onAuthStateChange((user) => setUser(user ?? null));
     * }, []);
     * ```
     */
    onAuthStateChange(cb) {
        this.authStateListeners.add(cb);
        return () => this.authStateListeners.delete(cb);
    }
    /**
     * Register callbacks for authentication events.
     */
    onAuthEvent(callbacks) {
        this.callbacks = { ...this.callbacks, ...callbacks };
    }
    clearAuthCallbacks() {
        this.callbacks = {};
    }
    // ─── waitForAuth ──────────────────────────────────────────────────────────────
    /**
     * Returns a Promise that resolves with the current user once `initAuth()` has completed.
     * If `initAuth()` has already run, resolves immediately.
     *
     * Useful in components that mount before auth state is known.
     *
     * @example
     * ```typescript
     * const user = await db.auth.waitForAuth();
     * if (user) showDashboard();
     * else showLogin();
     * ```
     */
    waitForAuth() {
        if (this.initAuthDone) {
            return Promise.resolve(this.user ?? null);
        }
        return new Promise((resolve) => {
            this.waitForAuthResolvers.push(resolve);
        });
    }
    // ─── initAuth ─────────────────────────────────────────────────────────────────
    /**
     * Initializes authentication by restoring the session from storage (or a provided token).
     *
     * - Checks JWT expiry — expired tokens are cleared without a server round-trip
     * - If user is not cached, fetches from server; on failure, clears session gracefully
     * - Returns `{ user, token, isAuthenticated }` so SSR can act on the result synchronously
     * - Fires `onAuthRestored` callback when complete (success or failure)
     *
     * @param token - Optional token to use directly (SSR: pass token from cookie/header).
     *                If omitted, reads from the configured storage adapter.
     *
     * @example
     * ```typescript
     * // Browser
     * const { isAuthenticated, user } = await db.auth.initAuth();
     *
     * // Next.js server component
     * const token = cookies().get('cocobase-token')?.value;
     * const { isAuthenticated, user } = await db.auth.initAuth(token);
     * ```
     */
    async initAuth(token) {
        const storedToken = token ?? this.storage.getItem(TOKEN_KEY);
        if (!storedToken) {
            this._finalizeInitAuth(undefined, undefined);
            return { user: undefined, token: undefined, isAuthenticated: false };
        }
        // Check JWT expiry before hitting the server
        const payload = this.decodeJwtPayload(storedToken);
        if (payload?.exp && Date.now() / 1000 > payload.exp) {
            this._clearSession();
            this._finalizeInitAuth(undefined, undefined);
            return { user: undefined, token: undefined, isAuthenticated: false };
        }
        this.token = storedToken;
        // Only persist to storage if we got the token from a param (SSR path)
        if (token) {
            this.storage.setItem(TOKEN_KEY, token);
        }
        // Try to restore cached user
        const cachedUserJson = this.storage.getItem(USER_KEY);
        if (cachedUserJson) {
            try {
                this.user = JSON.parse(cachedUserJson);
            }
            catch {
                this.user = undefined;
            }
        }
        // If no cached user, fetch from server
        if (!this.user) {
            try {
                await this.getCurrentUser();
            }
            catch (err) {
                // Only revoke the session when the server explicitly rejects the token.
                // Network errors, timeouts, and 5xx failures do NOT mean the token is
                // invalid — the user is still authenticated, the server is just unreachable.
                if (err instanceof UnauthorizedError || err instanceof ForbiddenError) {
                    this._clearSession();
                    this._finalizeInitAuth(undefined, undefined);
                    return { user: undefined, token: undefined, isAuthenticated: false };
                }
                // For all other errors (network, timeout, 5xx): keep the token in
                // memory so the user stays logged in.  user will be undefined until
                // the app makes a successful call to getCurrentUser() later.
            }
        }
        this._finalizeInitAuth(this.user, this.token);
        return {
            user: this.user,
            token: this.token,
            isAuthenticated: true,
        };
    }
    _finalizeInitAuth(user, token) {
        this.fireAuthStateChange();
        this.callbacks.onAuthRestored?.(user, !!token);
        this.initAuthDone = true;
        const resolvers = this.waitForAuthResolvers.splice(0);
        resolvers.forEach((resolve) => resolve(user ?? null));
    }
    _clearSession() {
        this.token = undefined;
        this.user = undefined;
        this.storage.removeItem(TOKEN_KEY);
        this.storage.removeItem(USER_KEY);
    }
    // ─── Login / Register ─────────────────────────────────────────────────────────
    /**
     * Authenticates a user with email and password.
     *
     * @example
     * ```typescript
     * const result = await db.auth.login({ email: 'user@example.com', password: 'pass' });
     * if (result.requires_2fa) {
     *   // prompt user for 2FA code, then call verify2FALogin
     * } else {
     *   console.log('Logged in as:', result.user?.email);
     * }
     * ```
     */
    async login(params) {
        const { email, password } = params;
        const response = await this.request("POST", `/auth-collections/login`, { email, password }, false);
        if (response.requires_2fa) {
            return { requires_2fa: true, message: response.message };
        }
        this.token = response.access_token;
        this.setToken(this.token);
        if (!response.user) {
            await this.getCurrentUser();
        }
        else {
            this.setUser(response.user);
        }
        if (this.user) {
            this.callbacks.onLogin?.(this.user, this.token);
            this.fireAuthStateChange();
        }
        return { requires_2fa: false, user: this.user };
    }
    /**
     * Registers a new user with email, password, and optional additional data.
     *
     * @example
     * ```typescript
     * const result = await db.auth.register({
     *   email: 'user@example.com',
     *   password: 'pass123',
     *   data: { username: 'johndoe' },
     * });
     * ```
     */
    async register(params) {
        const { email, password, data, roles, phone_number } = params;
        const response = await this.request("POST", `/auth-collections/signup`, { email, password, data, roles, phone_number }, false);
        if (response.requires_2fa) {
            return { requires_2fa: true, message: response.message };
        }
        this.token = response.access_token;
        this.setToken(this.token);
        if (!response.user) {
            await this.getCurrentUser();
        }
        else {
            this.setUser(response.user);
        }
        if (this.user) {
            this.callbacks.onRegister?.(this.user, this.token);
            this.fireAuthStateChange();
        }
        return { requires_2fa: false, user: this.user };
    }
    /**
     * Authenticates a user using Google Sign-In.
     *
     * @example
     * ```typescript
     * google.accounts.id.initialize({
     *   client_id: 'YOUR_GOOGLE_CLIENT_ID',
     *   callback: async ({ credential }) => {
     *     const user = await db.auth.loginWithGoogle({ idToken: credential });
     *   }
     * });
     * ```
     */
    async loginWithGoogle(params) {
        const { idToken, platform } = params;
        const response = await this.request("POST", "/auth-collections/google-verify", { id_token: idToken, platform }, false);
        this.token = response.access_token;
        this.setToken(response.access_token);
        this.setUser(response.user);
        this.callbacks.onLogin?.(response.user, response.access_token);
        this.fireAuthStateChange();
        return response.user;
    }
    /**
     * Authenticates a user using GitHub OAuth.
     *
     * @example
     * ```typescript
     * const code = new URLSearchParams(location.search).get('code');
     * const user = await db.auth.loginWithGithub({
     *   code,
     *   redirectUri: 'http://localhost:3000/auth/callback',
     * });
     * ```
     */
    async loginWithGithub(params) {
        const { code, redirectUri, platform } = params;
        const response = await this.request("POST", "/auth-collections/github-verify", { code, redirect_uri: redirectUri, platform }, false);
        this.token = response.access_token;
        this.setToken(response.access_token);
        this.setUser(response.user);
        this.callbacks.onLogin?.(response.user, response.access_token);
        this.fireAuthStateChange();
        return response.user;
    }
    /**
     * Register a new user with file uploads (avatar, cover photo, etc.)
     *
     * @example
     * ```typescript
     * await db.auth.registerWithFiles({
     *   email: 'john@example.com',
     *   password: 'pass123',
     *   data: { username: 'johndoe' },
     *   files: { avatar: avatarFile },
     * });
     * ```
     */
    async registerWithFiles(params) {
        const { email, password, data, roles, files } = params;
        const formData = new FormData();
        formData.append("data", JSON.stringify({ email, password, data, roles }));
        if (files) {
            for (const [fieldName, fileOrFiles] of Object.entries(files)) {
                if (Array.isArray(fileOrFiles)) {
                    fileOrFiles.forEach((file) => formData.append(fieldName, file));
                }
                else {
                    formData.append(fieldName, fileOrFiles);
                }
            }
        }
        const url = `${this.baseURL}/auth-collections/signup`;
        const res = await fetch(url, {
            method: "POST",
            headers: {
                ...(this.apiKey ? { "x-api-key": `${this.apiKey}` } : {}),
            },
            body: formData,
        });
        if (!res.ok) {
            const errorText = await res.text();
            throw new Error(`Registration failed: ${errorText}`);
        }
        const response = (await res.json());
        if (response.requires_2fa) {
            return { requires_2fa: true, message: response.message };
        }
        this.token = response.access_token;
        this.setToken(response.access_token);
        this.setUser(response.user);
        this.callbacks.onRegister?.(response.user, response.access_token);
        this.fireAuthStateChange();
        return { requires_2fa: false, user: response.user };
    }
    /**
     * Logs out the current user and clears all stored session data.
     *
     * @example
     * ```typescript
     * db.auth.logout();
     * ```
     */
    logout() {
        this._clearSession();
        this.callbacks.onLogout?.();
        this.callbacks.onTokenChange?.(undefined);
        this.fireAuthStateChange();
    }
    // ─── User management ──────────────────────────────────────────────────────────
    /**
     * Fetches the current authenticated user from the server.
     */
    async getCurrentUser() {
        if (!this.token) {
            throw new Error("User is not authenticated");
        }
        const user = await this.request("GET", `/auth-collections/user`);
        if (!user)
            throw new Error("Failed to fetch current user");
        this.user = user;
        this.setUser(user);
        return user;
    }
    /**
     * Updates the current user's profile.
     *
     * @example
     * ```typescript
     * await db.auth.updateUser({ data: { bio: 'Hello!' }, email: 'new@email.com' });
     * ```
     */
    async updateUser(params) {
        const { data, email, password } = params;
        if (!this.token)
            throw new Error("User is not authenticated");
        const body = {};
        if (data != null)
            body.data = mergeUserData(this.user?.data || {}, data);
        if (email != null)
            body.email = email;
        if (password != null)
            body.password = password;
        const user = await this.request("PATCH", "/auth-collections/user", body, false);
        this.user = user;
        this.setUser(user);
        this.callbacks.onUserUpdate?.(user);
        return user;
    }
    /**
     * Update current user with file uploads.
     *
     * @example
     * ```typescript
     * await db.auth.updateUserWithFiles({
     *   data: { bio: 'Updated' },
     *   files: { avatar: newAvatarFile },
     * });
     * ```
     */
    async updateUserWithFiles(params) {
        const { data, email, password, files } = params;
        if (!this.token)
            throw new Error("User is not authenticated");
        const formData = new FormData();
        const body = {};
        if (data != null)
            body.data = mergeUserData(this.user?.data || {}, data);
        if (email != null)
            body.email = email;
        if (password != null)
            body.password = password;
        if (Object.keys(body).length > 0) {
            formData.append("data", JSON.stringify(body));
        }
        if (files) {
            for (const [fieldName, fileOrFiles] of Object.entries(files)) {
                if (Array.isArray(fileOrFiles)) {
                    fileOrFiles.forEach((file) => formData.append(fieldName, file));
                }
                else {
                    formData.append(fieldName, fileOrFiles);
                }
            }
        }
        const url = `${this.baseURL}/auth-collections/user`;
        const res = await fetch(url, {
            method: "PATCH",
            headers: {
                ...(this.token ? { Authorization: `Bearer ${this.token}` } : {}),
            },
            body: formData,
        });
        if (!res.ok) {
            const errorText = await res.text();
            throw new Error(`User update failed: ${errorText}`);
        }
        const user = (await res.json());
        this.user = user;
        this.setUser(user);
        this.callbacks.onUserUpdate?.(user);
        return user;
    }
    /**
     * Checks if the current user has a specific role.
     *
     * @example
     * ```typescript
     * if (db.auth.hasRole('admin')) showAdminPanel();
     * ```
     */
    hasRole(role) {
        if (!this.user)
            throw new Error("User is not authenticated");
        return this.user.roles.includes(role);
    }
    /**
     * Lists users in the auth collection with optional filtering.
     *
     * @example
     * ```typescript
     * const users = await db.auth.listUsers({ limit: 20 });
     * ```
     */
    listUsers(query) {
        const query_str = buildFilterQuery(query);
        const url = `/auth-collections/users${query_str ? `?${query_str}` : ""}`;
        return this.request("GET", url);
    }
    /**
     * Gets a user by their ID.
     */
    getUserById(userId) {
        return this.request("GET", `/auth-collections/users/${userId}`);
    }
    requestPasswordReset(email) {
        return this.request("POST", "/auth-collections/forgot-password", { email }, false);
    }
    // ─── 2FA ─────────────────────────────────────────────────────────────────────
    /**
     * Enables Two-Factor Authentication for the current user.
     */
    enable2FA() {
        return this.request("POST", `/auth-collections/2fa/enable`, {}, false);
    }
    /**
     * Disables Two-Factor Authentication for the current user.
     */
    disable2FA() {
        return this.request("POST", `/auth-collections/2fa/disable`, {}, false);
    }
    /**
     * Sends a 2FA code to the user's email.
     */
    send2FACode(email) {
        return this.request("POST", `/auth-collections/2fa/send-code`, { email }, false);
    }
    /**
     * Completes login after 2FA verification.
     *
     * @example
     * ```typescript
     * const result = await db.auth.login({ email, password });
     * if (result.requires_2fa) {
     *   const user = await db.auth.verify2FALogin({ email, code: userEnteredCode });
     * }
     * ```
     */
    async verify2FALogin(params) {
        const { email, code } = params;
        const response = await this.request("POST", `/auth-collections/2fa/verify`, { email, code }, false);
        this.token = response.access_token;
        this.setToken(this.token);
        this.setUser(response.user);
        this.callbacks.onLogin?.(response.user, response.access_token);
        this.fireAuthStateChange();
        return response.user;
    }
    // ─── Email verification ───────────────────────────────────────────────────────
    requestEmailVerification() {
        return this.request("POST", `/auth-collections/verify-email/send`, {}, false);
    }
    verifyEmail(token) {
        return this.request("POST", `/auth-collections/verify-email/verify`, { token }, false);
    }
    resendVerificationEmail() {
        return this.request("POST", `/auth-collections/verify-email/resend`, {}, false);
    }
}
export default AuthHandler;
