// cocobase/dist/realtime/websockets.js
import { BASEURL } from "../utils/utils.js";
/**
 * Internal WebSocket wrapper with automatic reconnection and ping/pong keepalive.
 * Reconnect is suppressed once `close()` is called explicitly.
 */
class ReconnectingWebSocket {
    constructor(url, authData) {
        this.ws = null;
        this.reconnectDelay = 1000;
        this.maxReconnectDelay = 30000;
        this._closed = false;
        this._isReconnecting = false;
        this.url = url;
        this.authData = authData;
    }
    connect() {
        if (this._closed)
            return;
        this.ws = new WebSocket(this.url);
        this.ws.onopen = () => {
            const wasReconnecting = this._isReconnecting;
            this._isReconnecting = false;
            this.reconnectDelay = 1000;
            this.ws?.send(JSON.stringify(this.authData));
            this.pingInterval = globalThis.setInterval(() => {
                if (this.ws && this.ws.readyState === WebSocket.OPEN) {
                    this.ws.send(JSON.stringify({ type: "ping" }));
                }
            }, 30000);
            if (wasReconnecting) {
                this.reconnectHandler?.();
            }
        };
        this.ws.onclose = () => {
            if (this.pingInterval)
                clearInterval(this.pingInterval);
            if (this._closed)
                return;
            this._isReconnecting = true;
            setTimeout(() => {
                this.reconnectDelay = Math.min(this.reconnectDelay * 2, this.maxReconnectDelay);
                this.connect();
            }, this.reconnectDelay);
        };
        if (this.messageHandler) {
            this.ws.onmessage = this.messageHandler;
        }
    }
    send(msg) {
        if (this.ws && this.ws.readyState === WebSocket.OPEN) {
            this.ws.send(JSON.stringify(msg));
        }
    }
    close() {
        this._closed = true;
        if (this.ws)
            this.ws.close();
        if (this.pingInterval)
            clearInterval(this.pingInterval);
    }
    onMessage(cb) {
        this.messageHandler = cb;
        if (this.ws)
            this.ws.onmessage = cb;
    }
    onReconnect(cb) {
        this.reconnectHandler = cb;
    }
}
/**
 * Watches a collection for real-time create/update/delete events.
 *
 * @template T - The type of the document data
 *
 * @example
 * ```typescript
 * const watcher = db.realtime.collection<Post>('posts', { published: true });
 * watcher.connect();
 *
 * watcher.onCreate(({ document }) => {
 *   console.log('New post:', document?.data.title);
 * });
 *
 * watcher.onUpdate(({ document }) => {
 *   updatePost(document?.data);
 * });
 *
 * watcher.onDelete(({ document }) => {
 *   removePost(document?.id);
 * });
 *
 * watcher.onReconnect(() => {
 *   console.log('Reconnected — refreshing data...');
 * });
 *
 * // Cleanup
 * watcher.disconnect();
 * ```
 */
export class CollectionWatcher {
    constructor(collectionName, apiKey, filters) {
        this.url = `${BASEURL.replace(/^http/, "ws")}/realtime/collections/${collectionName}`;
        this.apiKey = apiKey;
        this.filters = filters;
    }
    connect() {
        this.reconnecting = new ReconnectingWebSocket(this.url, {
            api_key: this.apiKey,
            filters: this.filters,
        });
        this.reconnecting.connect();
        this.reconnecting.onMessage((ev) => {
            try {
                const data = JSON.parse(ev.data);
                const eventType = data.event;
                switch (eventType) {
                    case "connected":
                        this._onConnected?.(data);
                        break;
                    case "create":
                        this._onCreate?.(data);
                        break;
                    case "update":
                        this._onUpdate?.(data);
                        break;
                    case "delete":
                        this._onDelete?.(data);
                        break;
                    default:
                        break;
                }
            }
            catch (err) {
                this._onError?.({ error: err });
            }
        });
        this.reconnecting.onReconnect(() => {
            this._onReconnect?.();
        });
    }
    onConnected(cb) {
        this._onConnected = cb;
    }
    onCreate(cb) {
        this._onCreate = cb;
    }
    onUpdate(cb) {
        this._onUpdate = cb;
    }
    onDelete(cb) {
        this._onDelete = cb;
    }
    onError(cb) {
        this._onError = cb;
    }
    onClose(cb) {
        this._onClose = cb;
    }
    /**
     * Called when a dropped WebSocket connection is successfully re-established.
     * Use this to re-fetch data to catch any events missed during the disconnect.
     *
     * @example
     * ```typescript
     * watcher.onReconnect(async () => {
     *   const fresh = await db.listDocuments('posts');
     *   setPosts(fresh);
     * });
     * ```
     */
    onReconnect(cb) {
        this._onReconnect = cb;
    }
    disconnect() {
        this._onClose?.({});
        this.reconnecting?.close();
    }
}
/**
 * Broadcasts messages to all connected clients in a project.
 *
 * @template T - The type of message data
 *
 * @example
 * ```typescript
 * const broadcast = db.realtime.broadcast<{ text: string }>('user-123', 'Alice');
 * broadcast.connect();
 *
 * broadcast.onMessage((data) => {
 *   console.log('Received:', data);
 * });
 *
 * broadcast.send({ text: 'Hello everyone!' });
 * ```
 */
export class ProjectBroadcast {
    constructor(apiKey, userId, userName) {
        this.url = `${BASEURL.replace(/^http/, "ws")}/realtime/broadcast`;
        this.listeners = [];
        this.apiKey = apiKey;
        this.userId = userId;
        this.userName = userName;
    }
    connect() {
        this.reconnecting = new ReconnectingWebSocket(this.url, {
            api_key: this.apiKey,
            user_id: this.userId,
            user_name: this.userName,
        });
        this.reconnecting.connect();
        this.reconnecting.onMessage((ev) => {
            const data = JSON.parse(ev.data);
            if (data.type === "message") {
                this.listeners.forEach((cb) => cb(data));
            }
        });
    }
    send(data) {
        this.reconnecting?.send({ type: "message", data });
    }
    onMessage(cb) {
        this.listeners.push(cb);
    }
    /** Remove a previously registered message listener */
    offMessage(cb) {
        const idx = this.listeners.indexOf(cb);
        if (idx !== -1)
            this.listeners.splice(idx, 1);
    }
    disconnect() {
        this.reconnecting?.close();
    }
}
/**
 * Real-time room chat client.
 *
 * @example
 * ```typescript
 * const room = db.realtime.room('room-123', 'user-456', 'Alice');
 * room.create('My Chat Room');
 * // or room.join() if room already exists
 *
 * room.on('message', (data) => {
 *   appendMessage(data.content);
 * });
 *
 * room.sendMessage('Hello!');
 * room.leave();
 * ```
 */
export class RoomChat {
    constructor(roomId, apiKey, userId, userName) {
        this.listeners = {};
        this.url = `${BASEURL.replace(/^http/, "ws")}/realtime/rooms/${roomId}`;
        this.apiKey = apiKey;
        this.userId = userId;
        this.userName = userName;
    }
    create(roomTitle) {
        this._connect("create", roomTitle);
    }
    join() {
        this._connect("join");
    }
    _connect(action, roomTitle) {
        this.reconnecting = new ReconnectingWebSocket(this.url, {
            api_key: this.apiKey,
            user_id: this.userId,
            user_name: this.userName,
            action,
            room_title: roomTitle,
        });
        this.reconnecting.connect();
        this.reconnecting.onMessage((ev) => {
            const data = JSON.parse(ev.data);
            const eventType = data.event || data.type;
            (this.listeners[eventType] || []).forEach((cb) => cb(data));
        });
    }
    sendMessage(content) {
        this.reconnecting?.send({ type: "message", content });
    }
    on(event, cb) {
        if (!this.listeners[event])
            this.listeners[event] = [];
        this.listeners[event].push(cb);
        return () => {
            const arr = this.listeners[event];
            if (arr) {
                const idx = arr.indexOf(cb);
                if (idx !== -1)
                    arr.splice(idx, 1);
            }
        };
    }
    leave() {
        this.reconnecting?.close();
    }
}
export async function listRooms(apiKey) {
    const res = await fetch(`${BASEURL}/realtime/rooms`, {
        headers: { "X-API-Key": apiKey },
    });
    if (!res.ok) {
        const text = await res.text();
        throw new Error(`Failed to list rooms: ${text}`);
    }
    return res.json();
}
