// cocobase/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GameClient = exports.parseFilterKey = exports.buildFilterQuery = exports.uploadFile = exports.setToLocalStorage = exports.mergeUserData = exports.getFromLocalStorage = exports.Cocobase = void 0;
const core_js_1 = require("./core/core.js");
Object.defineProperty(exports, "Cocobase", { enumerable: true, get: function () { return core_js_1.Cocobase; } });
const utils_js_1 = require("./utils/utils.js");
Object.defineProperty(exports, "getFromLocalStorage", { enumerable: true, get: function () { return utils_js_1.getFromLocalStorage; } });
Object.defineProperty(exports, "mergeUserData", { enumerable: true, get: function () { return utils_js_1.mergeUserData; } });
Object.defineProperty(exports, "setToLocalStorage", { enumerable: true, get: function () { return utils_js_1.setToLocalStorage; } });
Object.defineProperty(exports, "buildFilterQuery", { enumerable: true, get: function () { return utils_js_1.buildFilterQuery; } });
Object.defineProperty(exports, "parseFilterKey", { enumerable: true, get: function () { return utils_js_1.parseFilterKey; } });
const file_js_1 = require("./core/file.js");
Object.defineProperty(exports, "uploadFile", { enumerable: true, get: function () { return file_js_1.uploadFile; } });
var multiplayer_1 = require("./realtime/multiplayer");
Object.defineProperty(exports, "GameClient", { enumerable: true, get: function () { return multiplayer_1.GameClient; } });
