// cocobase/dist/index.js
// ─── Main client ─────────────────────────────────────────────────────────────
export { Cocobase } from "./core/core.js";
// ─── Realtime classes (export so users can type variables) ────────────────────
export { CollectionWatcher, ProjectBroadcast, RoomChat } from "./realtime/websockets.js";
export { GameClient } from "./realtime/multiplayer.js";
// ─── Storage adapters ─────────────────────────────────────────────────────────
export { localStorageAdapter, memoryStorageAdapter, autoStorageAdapter, } from "./utils/storage.js";
export { cookieStorageAdapter } from "./utils/cookie.js";
// ─── Typed errors ─────────────────────────────────────────────────────────────
export { CocobaseError, NotFoundError, UnauthorizedError, ForbiddenError, RateLimitError, ValidationError, ServerError, TimeoutError, } from "./types/errors.js";
// ─── Utility functions ────────────────────────────────────────────────────────
export { getFromLocalStorage, mergeUserData, setToLocalStorage, buildFilterQuery, parseFilterKey, } from "./utils/utils.js";
export { uploadFile } from "./core/file.js";
