// cocobase/dist/cjs/realtime/websockets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoomChat = exports.ProjectBroadcast = exports.CollectionWatcher = void 0;
exports.listRooms = listRooms;
const utils_js_1 = require("../utils/utils.js");
class ReconnectingWebSocket {
    constructor(url, authData) {
        this.ws = null;
        this.reconnectDelay = 1000;
        this.maxReconnectDelay = 30000;
        this.url = url;
        this.authData = authData;
    }
    connect() {
        this.ws = new WebSocket(this.url);
        this.ws.onopen = () => {
            this.reconnectDelay = 1000;
            this.ws?.send(JSON.stringify(this.authData));
            // start ping/pong (use globalThis so Node environment works)
            this.pingInterval = globalThis.setInterval(() => {
                if (this.ws && this.ws.readyState === WebSocket.OPEN) {
                    this.ws.send(JSON.stringify({ type: "ping" }));
                }
            }, 30000);
        };
        this.ws.onclose = () => {
            if (this.pingInterval)
                clearInterval(this.pingInterval);
            setTimeout(() => {
                this.reconnectDelay = Math.min(this.reconnectDelay * 2, this.maxReconnectDelay);
                this.connect();
            }, this.reconnectDelay);
        };
    }
    send(msg) {
        if (this.ws && this.ws.readyState === WebSocket.OPEN) {
            this.ws.send(JSON.stringify(msg));
        }
    }
    close() {
        if (this.ws)
            this.ws.close();
        if (this.pingInterval)
            clearInterval(this.pingInterval);
    }
    onMessage(cb) {
        if (!this.ws)
            return;
        this.ws.onmessage = cb;
    }
}
class CollectionWatcher {
    constructor(collectionName, apiKey, filters) {
        this.url = `${utils_js_1.BASEURL.replace(/^http/, "ws")}/realtime/collections/${collectionName}`;
        this.apiKey = apiKey;
        this.filters = filters;
    }
    connect() {
        this.reconnecting = new ReconnectingWebSocket(this.url, {
            api_key: this.apiKey,
            filters: this.filters,
        });
        this.reconnecting.connect();
        this.reconnecting.onMessage((ev) => {
            try {
                const data = JSON.parse(ev.data);
                const eventType = data.event;
                switch (eventType) {
                    case "connected":
                        this._onConnected?.(data);
                        break;
                    case "create":
                        this._onCreate?.(data);
                        break;
                    case "update":
                        this._onUpdate?.(data);
                        break;
                    case "delete":
                        this._onDelete?.(data);
                        break;
                    default:
                        // unknown event
                        break;
                }
            }
            catch (err) {
                this._onError?.({ error: err });
            }
        });
    }
    onConnected(cb) {
        this._onConnected = cb;
    }
    onCreate(cb) {
        this._onCreate = cb;
    }
    onUpdate(cb) {
        this._onUpdate = cb;
    }
    onDelete(cb) {
        this._onDelete = cb;
    }
    onError(cb) {
        this._onError = cb;
    }
    onClose(cb) {
        this._onClose = cb;
    }
    disconnect() {
        this._onClose?.({});
        this.reconnecting?.close();
    }
}
exports.CollectionWatcher = CollectionWatcher;
class ProjectBroadcast {
    constructor(apiKey, userId, userName) {
        this.url = `${utils_js_1.BASEURL.replace(/^http/, "ws")}/realtime/broadcast`;
        this.listeners = [];
        this.apiKey = apiKey;
        this.userId = userId;
        this.userName = userName;
    }
    connect() {
        this.reconnecting = new ReconnectingWebSocket(this.url, {
            api_key: this.apiKey,
            user_id: this.userId,
            user_name: this.userName,
        });
        this.reconnecting.connect();
        this.reconnecting.onMessage((ev) => {
            const data = JSON.parse(ev.data);
            if (data.type === "message") {
                this.listeners.forEach((cb) => cb(data));
            }
        });
    }
    send(data) {
        this.reconnecting?.send({ type: "message", data });
    }
    onMessage(cb) {
        this.listeners.push(cb);
    }
    disconnect() {
        this.reconnecting?.close();
    }
}
exports.ProjectBroadcast = ProjectBroadcast;
class RoomChat {
    constructor(roomId, apiKey, userId, userName) {
        this.listeners = {};
        this.url = `${utils_js_1.BASEURL.replace(/^http/, "ws")}/realtime/rooms/${roomId}`;
        this.apiKey = apiKey;
        this.userId = userId;
        this.userName = userName;
    }
    create(roomTitle) {
        this._connect("create", roomTitle);
    }
    join() {
        this._connect("join");
    }
    _connect(action, roomTitle) {
        this.reconnecting = new ReconnectingWebSocket(this.url, {
            api_key: this.apiKey,
            user_id: this.userId,
            user_name: this.userName,
            action,
            room_title: roomTitle,
        });
        this.reconnecting.connect();
        this.reconnecting.onMessage((ev) => {
            const data = JSON.parse(ev.data);
            const eventType = data.event || data.type;
            (this.listeners[eventType] || []).forEach((cb) => cb(data));
        });
    }
    sendMessage(content) {
        this.reconnecting?.send({ type: "message", content });
    }
    on(event, cb) {
        if (!this.listeners[event])
            this.listeners[event] = [];
        this.listeners[event].push(cb);
    }
    leave() {
        this.reconnecting?.close();
    }
}
exports.RoomChat = RoomChat;
async function listRooms(apiKey) {
    const res = await fetch(`${utils_js_1.BASEURL}/realtime/rooms`, {
        headers: { "X-API-Key": apiKey },
    });
    if (!res.ok) {
        const text = await res.text();
        throw new Error(`Failed to list rooms: ${text}`);
    }
    return res.json();
}
